### ISSUE TITLE
Error in IntegrateData: starting vector near the null space

### ISSUE THREAD

**Issue by jeremymsimon** [satijalab/seurat#2570]
I'm working on a new dataset using v3.1.0, starting with loom objects that I convert and combine into one seurat object, split into a list of seurat objects (1 per replicate), scTransform, and then prep to integrate:

```
obj.features <- SelectIntegrationFeatures(object.list = all.list, nfeatures = 5000)
all.list <- PrepSCTIntegration(object.list = all.list, anchor.features = obj.features, verbose = TRUE)
obj.anchors <- FindIntegrationAnchors(object.list = all.list, normalization.method = "SCT", anchor.features = obj.features, verbose = TRUE)
obj.integrated <- IntegrateData(anchorset = obj.anchors, normalization.method = "SCT", verbose = TRUE)
```

Upon executing `IntegrateData` I get the following error:
```
Integrating data
Merging dataset 1 6 into 4 7 5 8 2
Extracting anchors for merged samples
Finding integration vectors
Warning in irlba(A = t(x = object), nv = npcs, ...) :
  fast code path error starting vector near the null space; re-trying with fastpath=FALSE.
Error in irlba(A = t(x = object), nv = npcs, ...) : 
  starting vector near the null space
```
This is a hard stop and the integrated object is not created.

This issue was also posted recently by @romanhaa on a different thread:
https://github.com/satijalab/seurat/issues/1720#issuecomment-577677825_

so it seems it's not just me. I have run other datasets recently too, so it's not an issue with my instance either. 

Any suggestions on how to navigate around this? Is it something that has been addressed in v3.1.1 or 3.1.2?

```
sessionInfo()
R version 3.5.2 (2018-12-20)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server 7.7 (Maipo)

Matrix products: default
BLAS/LAPACK: /nas/longleaf/apps/r/3.5.2/lib/libopenblas_haswellp-r0.3.5.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] loomR_0.2.1.9000  hdf5r_1.3.1.9000  R6_2.4.0          sctransform_0.2.0
[5] Seurat_3.1.0      reticulate_1.13  

loaded via a namespace (and not attached):
 [1] nlme_3.1-137        tsne_0.1-3          bitops_1.0-6       
 [4] bit64_0.9-7         RcppAnnoy_0.0.13    RColorBrewer_1.1-2 
 [7] httr_1.4.0          tools_3.5.2         irlba_2.3.3        
[10] KernSmooth_2.23-15  uwot_0.1.4          lazyeval_0.2.2     
[13] colorspace_1.4-1    npsurv_0.4-0        tidyselect_0.2.5   
[16] gridExtra_2.3       bit_1.1-14          compiler_3.5.2     
[19] plotly_4.9.0        caTools_1.17.1.1    scales_1.0.0       
[22] lmtest_0.9-36       ggridges_0.5.1      pbapply_1.4-0      
[25] stringr_1.4.0       digest_0.6.22       R.utils_2.8.0      
[28] pkgconfig_2.0.3     htmltools_0.4.0     bibtex_0.4.2       
[31] htmlwidgets_1.3     rlang_0.4.0         zoo_1.8-6          
[34] jsonlite_1.6        ica_1.0-2           gtools_3.8.1       
[37] dplyr_0.8.3         R.oo_1.22.0         magrittr_1.5       
[40] Matrix_1.2-17       Rcpp_1.0.2          munsell_0.5.0      
[43] ape_5.2             R.methodsS3_1.7.1   stringi_1.4.3      
[46] gbRd_0.4-11         MASS_7.3-51.1       gplots_3.0.1.1     
[49] Rtsne_0.15          plyr_1.8.4          grid_3.5.2         
[52] parallel_3.5.2      gdata_2.18.0        listenv_0.7.0      
[55] ggrepel_0.8.1       crayon_1.3.4        lattice_0.20-38    
[58] cowplot_1.0.0       splines_3.5.2       SDMTools_1.1-221   
[61] pillar_1.4.2        igraph_1.2.4        future.apply_1.3.0 
[64] reshape2_1.4.3      codetools_0.2-16    leiden_0.3.1       
[67] glue_1.3.1          lsei_1.2-0          metap_1.1          
[70] data.table_1.12.8   RcppParallel_4.4.2  png_0.1-7          
[73] Rdpack_0.10-1       gtable_0.3.0        RANN_2.6.1         
[76] purrr_0.3.0         tidyr_0.8.2         future_1.14.0      
[79] assertthat_0.2.1    ggplot2_3.2.1       rsvd_1.0.2         
[82] survival_2.44-1.1   viridisLite_0.3.0   tibble_2.1.3       
[85] cluster_2.1.0       globals_0.12.4      fitdistrplus_1.0-14
[88] ROCR_1.0-7         
```

**Comment by romanhaa** [comment]
I just added a comment to the post you linked: https://github.com/satijalab/seurat/issues/1720#issuecomment-581571671. Might work in your case as well, even though it seems that you don't have that many data sets to integrate. Anyway, might be worth a try.

**Comment by jeremymsimon** [comment]
Thanks @romanhaa. Mine is roughly 25K cells to start with, and I'm not getting any memory-related errors from what I can tell. Given the biology involved, I'd *really* love to not specify a reference sample though unless it is absolutely necessary.

**Comment by romanhaa** [comment]
I never got any memory-related errors either, and I'm not sure it ever was a problem. I only know that at some point down-sampling helped. 25k cells however really isn't that much.

Btw, from my experience using reference samples or not didn't make any difference, it's just that in my case that's what made sense. I tried not specifying some samples as reference and just integrating them all equally but ran into the same error as before.

**Comment by jeremymsimon** [comment]
In our experiment, there may be cells that are completely unique to one condition or the other, so it seems safer (at least to me) to not specify a reference - I don't want those cells to be missed somehow

**Comment by romanhaa** [comment]
Sorry, what I said was not clear. When I said that specifying reference samples or not doesn't make any difference in my experience, I meant that regarding whether the `IntegrateData()` function will throw an error or not. Of course, if you don't have any actual references, I wouldn't recommend specifying any sample as a reference.

**Comment by jeremymsimon** [comment]
I've tried reducing the number of features from 5000 to 3000, and tried specifying more memory on our supercomputer. Neither of these got around this error. It appears as though it's using ~75GB of memory of the 200GB I allotted, and gets 7/8 of the way through the integration. Really not sure what's going on here, but would really love to hear some ideas on a workaround

**Comment by jeremymsimon** [comment]
Anyone else have any ideas on this? I tried allotting 200GB of RAM and it still fails, but a `gc()` reports that it's only using ~75GB before it dies. I'm very much stuck on this though so any help would be appreciated!

**Comment by satijalab** [comment]
Thanks Jeremy. The code here looks correct - so we'd like to check this out in more detail. Can you either attach on object or e-mail a dropbox download link to tstuart@nygenome.org? Please try to reproduce the error with the minimal number of cells to make it easier for us to debug and share.

**Comment by jeremymsimon** [comment]
Thanks- I just shared the file via dropbox and shared my entire code block to help you reproduce the error. Let me know if that didn't come through or if you have any issues or questions

**Comment by timoast** [comment]
Thanks @jeremymsimon, I have received the data and am looking into it now

**Comment by timoast** [comment]
@jeremymsimon I don't think you've sent the necessary object, can you please send the object called `all.list` in your original code, ie:

```r
obj.features <- SelectIntegrationFeatures(object.list = all.list, nfeatures = 5000)
all.list <- PrepSCTIntegration(object.list = all.list, anchor.features = obj.features, verbose = TRUE)
obj.anchors <- FindIntegrationAnchors(object.list = all.list, normalization.method = "SCT", anchor.features = obj.features, verbose = TRUE)
obj.integrated <- IntegrateData(anchorset = obj.anchors, normalization.method = "SCT", verbose = TRUE)
```

Please run:

```r
saveRDS(object = all.list, file = "./object.rds")
```

and send `object.rds`

**Comment by jeremymsimon** [comment]
Shared `object.rds` via Dropbox

**Comment by timoast** [comment]
@jeremymsimon I wasn't able to reproduce your error. I ran the following code with the dataset you sent and it completed sucessfully:

```r
all.list <- readRDS('/home/stuartt/data/object.rds')

top1.features <- SelectIntegrationFeatures(object.list = all.list, nfeatures = 3000)
all.list <- PrepSCTIntegration(object.list = all.list, anchor.features = top1.features, verbose = TRUE)
top1.anchors <- FindIntegrationAnchors(object.list = all.list, normalization.method = "SCT", anchor.features = top1.features, verbose = TRUE)
top1.integrated <- IntegrateData(anchorset = top1.anchors, normalization.method = "SCT", verbose = TRUE)
```

Not sure why you're seeing that error but I'd suggest updating to the latest version from github and try again

**Comment by jeremymsimon** [comment]
Yes this does seem to be fixed as of version 3.1.4. Do you foresee any issues updating Seurat 3.1.0 objects to 3.1.4 or should that be a relatively minor update?

**Comment by timoast** [comment]
No that's a minor update
